codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/ServiceConfigGenerator.kt

```diff
@@ -6,31 +6,141 @@
 package software.amazon.smithy.rust.codegen.server.smithy.generators
 
 import software.amazon.smithy.rust.codegen.core.rustlang.RustWriter
+import software.amazon.smithy.rust.codegen.core.rustlang.Writable
+import software.amazon.smithy.rust.codegen.core.rustlang.conditionalBlock
+import software.amazon.smithy.rust.codegen.core.rustlang.docs
+import software.amazon.smithy.rust.codegen.core.rustlang.join
+import software.amazon.smithy.rust.codegen.core.rustlang.rust
+import software.amazon.smithy.rust.codegen.core.rustlang.rustBlockTemplate
 import software.amazon.smithy.rust.codegen.core.rustlang.rustTemplate
+import software.amazon.smithy.rust.codegen.core.rustlang.writable
 import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType
+import software.amazon.smithy.rust.codegen.core.smithy.RuntimeType.Companion.preludeScope
 import software.amazon.smithy.rust.codegen.core.util.toPascalCase
 import software.amazon.smithy.rust.codegen.server.smithy.ServerCargoDependency
 import software.amazon.smithy.rust.codegen.server.smithy.ServerCodegenContext
 
+fun List<ConfigMethod>.isBuilderFallible() = this.any { it.isRequired }
+
+/**
+ * Contains all data necessary to render a method on the config builder object to apply arbitrary layers, HTTP plugins,
+ * and model plugins.
+ */
+data class ConfigMethod(
+    /** The name of the method. **/
+    val name: String,
+    /** The Rust docs for the method. **/
+    val docs: String,
+    /** The parameters of the method. **/
+    val params: List<Binding>,
+    /** In case the method is fallible, the error type it returns. **/
+    val errorType: RuntimeType?,
+    /** The code block inside the method. **/
+    val initializer: Initializer,
+    /** Whether the user must invoke the method or not. **/
+    val isRequired: Boolean,
```

Comment:
> If `isRequired` is set to true and the user fails to set it, will the `errorType` parameter be returned?"

They're independent, they have nothing to do with each other. `isRequired` tracks whether the user must invoke the method or not. `errorType` is the error type returned when you invoke the method, and the code inside it can fail.

> How can I create a `ConfigMethod` that stores data in the `Config` structure, which can later be used by `$serviceName::builder()`?

That's not allowed. The methods can only perform checks and prepare the layers and plugins, which will be applied before they return.